src/mathf/geometry/multi-dimensional-vector.ts

```diff
@@ -0,0 +1,191 @@
+import { arrayf } from '../../arrayf/arrayf';
+import { mathf } from '../mathf';
+
+/**
+ * Class for handling vectors of arbitrary dimensions
+ */
+export class MultiDimensionalVector {
```

Comment:
just a general note: it's really hard for me to follow this code mainly because everything is so heavily abstracted and has generic typings, etc.. i would suggest maybe looking at similar code and seeing if this code can be simplified.

for example, instead of accepting generic types, is it at all appropriate to accept explicit types?

or if `Dimensions2d` is the only implementation of `MultiDimensionalVector` right now, i'd suggest consolidating everything into the 2d class until we have a need for this base class.

compare to:

https://github.com/maxkueng/victor/blob/master/index.js

which is a little easier for me to read and follow.

Reply:
In my head it was a trade-off between ease-of-use working within the abstract class' code, but gaining ease-of-use working with the instances of its extended classes and narrowing maintenance of generic behaviour to a single point.

By being this generic it:
- Supports 2d/3d vectors, areas, volumes and colors
- Limits maintenance of common functions to a single point
- Limits testing of common functions to a single point
- Provides a consistent API and behaviour across vectors

3d vectors and Colors are still sitting in Toolbox waiting to be ported or consolidated with the existing Yano classes, since their need wasn't as urgent for current work.

The cons are:
- Readability of generic/abstracts
- Full benefit would depend on eventually having the existing Vector/Color classes in Yano extend this base

If this is acceptable we can leave it as is.

We can simplify a bit if `: MultiDimensionalVector` is used instead of `: <T extends MultiDimensionalVector>`, casting returned values to the desired type.

We can remove this class and duplicate functionality as needed into child classes, accepting potentially different behaviour per vector. API could be kept consistent via an Interface.

We could use an Interface to keep the API consistent and remove the generic class, replacing it with a decorator that adds the duplicated functionality to classes as needed.

We could use an interface to keep the API consistent, remove the generic class, treat instances as immutable data types and use static functions on classes for manipulations. Essentially having a vectorf type class.

So we can:
- Leave as is, add a note that this is not intended to be used directly
- Replaces instances of `T extends MultiDimensionalVector` with `MultiDimensionalVector` and cast extended types as needed.
- Remove abstract class, duplicate code into explicit classes
- Remove abstract class, create decorator
- Remove abstract class, treat instances as immutable data types, use static functions for generic behaviour.

Reply:
i think for me personally i would love for this module to be simplified somehow. as someone not super familiar with this code, i would have a hard time figuring out how to instantiate a vector and to call its appropriate functions.

i wish i had better suggestions, but i haven't used vectors too much. mind checking closure library and the aforementioned `victor` lib to see if there are potentially better ways to structure this?

https://github.com/google/closure-library/tree/master/closure/goog

if you're pressed for time and need to get this in, please don't let me block this review.

Reply:
Angus - in light of trying to reduce abstraction and keeping it simple, can you use yano.vector instead?  

IMO, I think the mix of static and non static classes tends to be an overhead and it's just simpler to keep that straight-forward instead of doing what has been done here: https://github.com/grow/yano-js/blob/821fea82a320f1b13a0c4c3e205a116ab3a36aa9/src/mathf/geometry/multi-dimensional-vector.ts#L75 but perhaps there was a reason for this.

How often do you use Dimensions2D?  it seems to basically be just a vector2D except calling x and y  - width and height instead so I didn't really see a strong need for that but correct me if I'm wrong.

Reply:
The closure library does what I was hoping to avoid, which is duplicating code and creating multiple maintenance and test points.

It's not intended to be used directly, but as a base for Vector2d Dimensions2d etc.

It would be used like:

```
new MultiDimensionalVector(1, 0, 7).add(new MultiDimensionalVector(2, 3, -1));

// Returning equivalent of new MultiDimensionalVector(3, 3, 6)
```

But given more explicit values seems to be the preferred method I'll remove this class and re-create its functionality in its sub-classes as needed.